### ISSUE TITLE
Implement some functions from other decimal libraries

### ISSUE THREAD

**Issue by matejsp** [quagmt/udecimal#22]
I really like the fallback used by this library by using int64 for making it faster as fallback.
What I was try to do was to create shopspring wrapper around this library. 
This way we can benchmark if this library works and is faster.

But I am missing couple of functions:

1. General functions:
```
// IntPart returns the integer component of the decimal.
func (d Decimal) IntPart() int64

// BigInt returns integer component of the decimal as a BigInt.
func (d Decimal) BigInt() *big.Int

// Pow returns d to the power of d2.
func (d Decimal) Pow(d2 Decimal) Decimal

// Mod returns d % d2.
func (d Decimal) Mod(d2 Decimal) Decimal

// Float64 that returns error if not exact
func (d Decimal) Float64() (float64, error) {
```

2. I am missing all the rounding functions.
Supported are just 3 (A bit unconventional naming): RoundUp => RoundHTZ, RoundDown => RoundHAZ, RoundBank

```
// RoundCeil rounds the decimal towards +infinity.
func (d Decimal) RoundCeil(places int32) Decimal {

// RoundFloor rounds the decimal towards -infinity.
func (d Decimal) RoundFloor(places int32) Decimal {
```

All that are supported are Ceil and Floor without precision.

3. Unable to parse exponent string format:
panic: invalid format: can't parse '1e-8' to Decimal
panic: invalid format: can't parse '1e6' to Decimal

**Comment by quagmt** [comment]
What do you mean by `create shopspring wrapper around this library`?

I want to clarify that this library is designed to provide a more efficient way to do finanacial calculations. Making it fully compatible with shopsping/decimal is never a goal so you'll expect some functions in shopspring are missing or naming can be different. I'm open to add new features if there are strong use cases, which I don't really see in those you mentioned.

For correctness, all methods are already fuzzed in [fuzz_test.go](https://github.com/quagmt/udecimal/blob/master/fuzz_test.go). The results are also cross-checked with shopspring/decimal. If you have any doubt, I would be very appreciate if  you can try running these fuzz tests yourself locally and report any abnormal results.

**Comment by matejsp** [comment]
I don't expect library as drop in replacement (would be interesting idea though :D).

I was just pointing out the functionality that are missing (not function naming).
- int part of decimal, 
- big intpart of decimal, 
- power function with decimal, 
- mod function 
- round ceil (with precision like RoundHTZ)
- round floor (with precision like RoundHTZ)

We currently use shopspring library. To test how you library behaves inside our code would be very hard without doing complete migrating That's why I created decimal struct that implements all the methods from shopspring using your library. I like the potential performance improvement. But need to achieve feature parity first.

```
type Decimal struct {
	v udecimal.Decimal
}

func (d Decimal) Mul(d2 Decimal) Decimal {
	v := d.v.Mul(d2.v)
	if DoValidateMulDiv {
		Validate(v, d.ToShopDecimal().Mul(d2.ToShopDecimal()), d.v.String(), "*", d2.v.String())
	}
	return Decimal{v: v}
}

func (d Decimal) GreaterThan(d2 Decimal) bool {
	return d.v.Cmp(d2.v) > 0
}

func (d Decimal) IntPart() int64 {
	whole := int64(d.v.InexactFloat64())
	if DoValidate {
		ValidateStr(strconv.Itoa(int(whole)), strconv.Itoa(int(d.ToShopDecimal().IntPart())))
	}
	return whole
}

func (d Decimal) RoundFloor(places int32) Decimal {
	var v udecimal.Decimal
	if places == 0 {
		v = d.v.Floor()
	} else {
		v, _ = d.v.Mul(Ten.v.PowInt(int(places))).Floor().Div(Ten.v.PowInt(int(places)))
	}
	if DoValidate {
		Validate(v, d.ToShopDecimal().RoundFloor(places))
	}
	return Decimal{v: v}
}

```

Then I use go.mod replace with wrapped udecimal library. Basically running our app with your library using shopspring api.

**Comment by matejsp** [comment]
Maybe one difference that I noticed that shopspring, python do not limit digit precision for multiplication. udecimal does to 19 even for multiplication.

For division python limits to 23, shopspring to 16 and udecimal to 19.

This library has max and defaut 19. Would it be possible to increase the precision (at least max) to 23?

```
package decimal

import (
	shop "github.com/shopspring/decimal"
	udecimal "github.com/quagmt/udecimal"
	"github.com/stretchr/testify/assert"
	"testing"
)

func TestPrecisions(t *testing.T) {

	t.Run("Div", func(t *testing.T) {
		// panic: -33333.3333333333333333333 != -33333.3333333333333333, [-50000 / 1.5]
		// -33333.3333 3333 3333 3333 333 !=
		// -33333.3333 3333 3333 3333

		a1 := shop.RequireFromString("-50000")
		b1 := shop.RequireFromString("1.5")
		c1 := a1.Div(b1)

		udecimal.SetDefaultPrecision(16)
		a2 := udecimal.MustParse("-50000")
		b2 := udecimal.MustParse("1.5")
		c2, _ := a2.Div(b2)

		udecimal.SetDefaultPrecision(19)
		a3 := udecimal.MustParse("-50000")
		b3 := udecimal.MustParse("1.5")
		c3, _ := a3.Div(b3)

		assert.Equal(t, c1.String(), c2.String())
		assert.Equal(t, c1.String(), c3.String())
	})

	t.Run("Mul", func(t *testing.T) {
		// panic: -33343.3333333333333332 != -33343.33333333333333329999, [-33333.3333333333333333 * 1.0003]

		a1 := shop.RequireFromString("-33343.3333333333333333")
		b1 := shop.RequireFromString("1.0003")
		c1 := a1.Mul(b1)

		udecimal.SetDefaultPrecision(16)
		a2 := udecimal.MustParse("-33343.3333333333333333")
		b2 := udecimal.MustParse("1.0003")
		c2 := a2.Mul(b2)

		udecimal.SetDefaultPrecision(19)
		a3 := udecimal.MustParse("-33343.3333333333333333")
		b3 := udecimal.MustParse("1.0003")
		c3 := a3.Mul(b3)

		assert.Equal(t, c1.String(), c2.String())
		assert.Equal(t, c1.String(), c3.String())
	})

}
```

But what is different is that in python I get:
```
import decimal

a = decimal.Decimal("-33333.3333333333333333")
b = decimal.Decimal("1.0003")
print(a * b) # -33343.33333333333333329999

a = decimal.Decimal("-33333.3333333333333333")
b = decimal.Decimal("1.000003")
print(a * b)

a = decimal.Decimal("-50000")
b = decimal.Decimal("1.5")
print(a / b)
```

```
-33333.3333333333333333 * 1.0003=

python decimal: -33343.33333333333333329999
shopspring:     -33343.33333333333333329999
udecimal:       -33343.3333333333333332999
````
or 

```
-33333.3333333333333333 * 1.0000003=
python decimal: -33333.4333333333333332999999
shopspring:     -33343.4333633333333332999999
udecimal:       -33343.4333633333333332999
```

**Comment by quagmt** [comment]
I suggest you fork this repo and add you own modification to suit your use cases because it seems like you want to add these methods to make it compatible with shopspring/decimal and avoid code migration while testing. As I mentioned is not a goal for this library.
That being said, I'm still interested in your actual use cases for these methods because they seem odd to me when dealing with financial numbers. For example:
- IntPart() int64 returns an undefined value if the integer part doesn't fit into int64 and also doesn't return an error. In which situation do you need this function, and how do you use it correctly?
- BigInt() big.Int seems more reasonable, but using big.Int directly will definitely hurt performance and should be avoided as much as possible.
- Pow(v decimal) is also not common when dealing with financial numbers, which is why I only implemented PowInt. Do you use decimal as a power, such as (2.5).Pow(1.25), or is the power always an integer?
- Mod sounds a bit odd to me. I checked shopspring/decimal and it returns q as integer and r as decimal. What's your use case for this method?

For the rounding methods, there are some concerns I want to clarify:
- RoundUp and RoundDown are actually RoundHalfAwayFromZero and RoundHalfTowardZero with different names. For me, using different names makes it more confusing and increases cognitive complexity because I have to remember that `RoundUp = Half Away From Zero`. There's also another rounding method called [RoundHalfUp](https://en.wikipedia.org/wiki/Rounding#Rounding_half_up), which makes it even more confusing.
- For RoundCeil (or RoundFloor), the behavior is different between positive and negative numbers. For example:
```go
	a := decimal.RequireFromString("1.1234").RoundCeil(3) // 1.1234 --> 1.124
	b := decimal.RequireFromString("-1.1234").RoundCeil(3) // -1.1234 --> 1.123
```

Is this your expected behavior, or do you only care about positive numbers? All implemented rounding methods are symmetrical so far; I want to maintain that behavior for new rounding methods. It would be something like RoundTowardsInf, for example:
```go
     	a := MustParse("1.1234").RoundTowardsInf(3)  // 1.1234 -> 1.124
	b := MustParse("-1.1234").RoundTowardsInf(3) // -1.1234 -> -1.124
```

**Comment by quagmt** [comment]
> Would it be possible to increase the precision (at least max) to 23?

It would involve a lot of rework. Can you explain why you want to have 23 digits? Is it just to follow the Python implementation? It seems you are fine with the 16 digits from the shopspring Div method tho.

Also, the choice of 19 decimal digits is not random. I agree that more digits are generally better. However, 19 digits provide certain advantages for different methods in this library. For example:
- The whole number (if not overflowing) is stored in uint128. The maximum number of digits that uint128 can hold is 39, with 19 digits for the decimal places and 20 digits for the integer part. If we increase the decimal places, the data range of decimal values that can be stored in uint128 will decrease.
- 19 digits also makes the fractional part fit into a uint64, which helps increase performance when converting the decimal to a string (you can check the stringU128 method). The string method is usually called when writing to a database, so this will help improve database write performance.
- 19 digits also impact the Div method. Before performing division, there's an adjustment for the dividend to ensure the result always has 19 digits after the decimal point. The adjustment involves multiplying the dividend by `pow10[defaultPrec + divisor.Prec] <= pow10[38] if prec <=19`.  `pow10[38]` is also the maximum power of 10 that can be stored in uint128. If we increase the maximum precision, for example to 23, then we might need to calculate pow10[46], which doesn't fit into uint128 and will fall back to big.Int API. That would make the entire Div function less performant.

**Comment by matejsp** [comment]
> IntPart() int64 returns an undefined value if the integer part doesn't fit into int64 and also doesn't return an error. In which situation do you need this function, and how do you use it correctly?

Well based on the context. Let's be honest that IntPart in its limitation can accomodate up to 9223372036854775807 (whole precision is 18 numbers). Its more than enough for 99.99% financial cases. But you could return func IntPart() (int64, error) to be safer.

> Pow(v decimal) is also not common when dealing with financial numbers, which is why I only implemented PowInt. Do you use decimal as a power, such as (2.5).Pow(1.25), or is the power always an integer?

We actually always use it with integer. But the tricky part is missing func IntPart :D Because you cannot do Pow from Decimal directly. We calculate pow factor by taking the int part of calculated decimal number.
```
baseCounterRatio.RoundBank(int32(ratioValueDecimal.IntPart())), nil
```

> RoundUp and RoundDown are actually RoundHalfAwayFromZero and RoundHalfTowardZero with different names.
https://en.wikipedia.org/wiki/Rounding#Types_of_rounding

Hm should be: RoundUp is Round Away From Zero and not Round Half Away From Zero.
Having the same issue with cognitive complexity. I am a bit torn about standard names in other languages/libraries and ones in this library:
https://github.com/rin01/decnum/blob/master/mydecquad.go#L208-L215
https://docs.oracle.com/javase/8/docs/api/java/math/RoundingMode.html
https://docs.python.org/3/library/decimal.html#rounding-modes
https://support.microsoft.com/en-us/office/roundup-function-f8bc9b23-e795-47db-8703-db171d0c42a7

I don't want to question function names in your library but to support all the standard rounding mechanisms (8 standard ones vs only 3 here). Can be some forms of RoundTZ, RoundTI, etc ...

> Is this your expected behavior, or do you only care about positive numbers.

I care about positive AND negative rounding that is standard compatible with other implementations, yes including excel.

> Mod sounds a bit odd to me. I checked shopspring/decimal and it returns q as integer and r as decimal. What's your use case for this method?

We use it in one of our financial risk algorithms where we search for remainder of decimal division. Like 5.1 mod 1.5 = 0.6

> It would involve a lot of rework. Can you explain why you want to have 23 digits? Is it just to follow the Python implementation? It seems you are fine with the 16 digits from the shopspring Div method tho.

I understand. It is mostly for compatibility when migrating and optimizing existing code. We can change precision but we need to fix a lot of test cases because we test to maximum precision supported by lib in intermediate results and round only final result.

> I suggest you fork this repo and add you own modification to suit your use cases because it seems like you want to add these methods to make it compatible with shopspring/decimal and avoid code migration while testing. As I mentioned, that is not a goal for this library

As I mentioned this is not the feature I am looking from this library. I care about the mathematical features that we use in financial calculations.

Perhaps forking would be easier way and probably better direction for us. However I always prefer collaboration if it can result in a better library. At least on the parts (math functionality) that we can agree on.

We are using decimal libraries for financial risk, liabilities, liquidation and exposures in futures trading. This includes a lot of backtracking formulas like WMA, trading indicators, price back tracking etc. Some algorithms and test cases are migrated from python, java and some even from excel.

**Comment by quagmt** [comment]
> Hm should be: RoundUp is Round Away From Zero and not Round Half Away From Zero.

Oops. My bad. Thanks for the correction.

- I'm okay to add the `IntPart()` function but it will be in form of this `IntPart() (int,error)`. It returns error when the integer part doesn't fit int32 instead of int64. The reason is that `PowInt` is accepting an `int`. Returning int64 from `IntPart()` then convert it to int might not work on 32-bit machines for some cases. It's kinda unelegant for now but have to make it that way to avoid unexpected behavior. `PowInt` can technically receives `int64` as input but it's a breaking change. I would like to have it for maybe v2 in the future.
- For `Mod` function, I checked C language and it also has `fmod` which is similar to this. Seems like a reasonable feature to me. I'll add it.
- For rounding, I still prefer all methods to be symmetrical. I think it's only missing `RoundToInf` (towards infinity) now. Here's the feature parity for rounding:
```
other libs         |    udecimal
---------------------------------
Round_Up           | RoundToInf
Round_Down         | Trunc
Round_Half_Even    | RoundBank
Round_Half_Up      | RoundHAZ
Round_Half_Down    | RoundHTZ
Round_Ceil         | not_supported
Round_Floor        | not_supported
```
How common is it when you use `Round_Ceil` or `Round_Floor`? Can you use `RoundToInf` instead?

**Comment by matejsp** [comment]
That's would be awesome. 

I checked in our code. We use RoundCeil and RoundFloor in around 20 and other 30 places. While RoundUp and RoundDown in more than 50 occurrences. RoundBank around 50. But we don't use HalfUp and HalfDown. 

Changing the rounding is of course possible but adds complexity because you need to handle positive and negative numbers using reduced number of operations and increases complexity since calculations and roundings are based on product specifications. 
We could use helper functions. But I am missing rounding to negative inf (trunc for positive, ? for negative).

RoundUp is not RoundToInf but RoundAZ (away from zero). It differs for negative numbers:
```
RoundUp(-1.1) == -2 (to -inf if negative, +inf if positive ... away from zero)
RoundCeil(-1.1) == -1 (negative/positive number is rounded to +inf)
RoundToInf(-1.1) == -1 (negative number is rounded (to +inf)
```

```
RoundCeiling  RoundingMode = C.DEC_ROUND_CEILING   // Round towards +Infinity.
RoundDown     RoundingMode = C.DEC_ROUND_DOWN      // Round towards 0 (truncation).
RoundFloor    RoundingMode = C.DEC_ROUND_FLOOR     // Round towards –Infinity.
RoundHalfDown RoundingMode = C.DEC_ROUND_HALF_DOWN // Round to nearest; if equidistant, round down.
RoundHalfEven RoundingMode = C.DEC_ROUND_HALF_EVEN // Round to nearest; if equidistant, round so that the final digit is even.
RoundHalfUp   RoundingMode = C.DEC_ROUND_HALF_UP   // Round to nearest; if equidistant, round up.
RoundUp       RoundingMode = C.DEC_ROUND_UP        // Round away from 0.
Round05Up     RoundingMode = C.DEC_ROUND_05UP      // The same as RoundUp, except that rounding up only occurs if the digit to be rounded up is 0 or 5 and after Overflow the result is the same as for RoundDown.
RoundDefault  RoundingMode = RoundHalfEven         // The same as RoundHalfEven.
```

Based on that:
```
Round_Up           | RoundAZ -> away from zero
Round_Down         | RoundTZ -> to zero -> Trunc also works, but I like parity for rounding too.
Round_Half_Even    | RoundBank
Round_Half_Up      | RoundHAZ
Round_Half_Down    | RoundHTZ
Round_Ceil         | RoundToInf -> round to positive infinity
Round_Floor        | RoundToNegInf -> round to negative infinitiy
```

**Comment by quagmt** [comment]
Can you explain why you use `RoundUp` in some places and `RoundCeil` in other places? I want to make rounding with this library as simple as possible by not adding too many rounding methods that can cause confusion when choosing which one to use, and keep the behavior similar for both positive and negative numbers. `RoundToInf` means round to infinity, so if it will round up to +inf if positive and to -inf if negative. I already have an example for it
```go
     	a := MustParse("1.1234").RoundTowardsInf(3)  // 1.1234 -> 1.124
	b := MustParse("-1.1234").RoundTowardsInf(3) // -1.1234 -> -1.124

        // mean while round ceil
        a := decimal.RequireFromString("1.1234").RoundCeil(3) // 1.1234 --> 1.124
	b := decimal.RequireFromString("-1.1234").RoundCeil(3) // -1.1234 --> 1.123
```

Also, for `IntPart`, actually I think it's better if `IntPart` returns a decimal, and there's a `Pow(e decimal)`, so usage is very simple and straight forward. For example your use case will look something like this
```go
      a.Pow(ratio.Trunc(0))
```
However, Pow by decimal is complicated so it'll probably take some times.


Btw, I double checked your example when describing Pow but it's actually RoundBank, is there any mistakes?
```go
baseCounterRatio.RoundBank(int32(ratioValueDecimal.IntPart())), nil
```

**Comment by matejsp** [comment]
```
Can you explain why you use RoundUp in some places and RoundCeil in other places?
```

For example we use RoundCeil to round the fees that user needs to pay per perp trade. Fees are negative for maker and positive for takes. RoundCeil is making sure that we round the fees towards inf because you want to give little less and take a bit more. This is important because net fee should never be bellow 0. For funding rate that can also be negative or positive we use RoundFloor because you want to take a bit more and give a bit more (net total should be very close to zero but never bellow 0). Rounding error bits goes to insurance fund. RoundUp is used when calculating short (- amount) and long (+ amount) exposure because you want to always round up (towards +inf).

```
Btw, I double checked your example when describing Pow but it's actually RoundBank, is there any mistakes?
```

Yes, we have a lot of formulas :) another with Pow that we use:
if collateralAvailableForLiquidation.GreaterThanOrEqual(decimal.NewFromInt(10).Pow(spotMarket.BasePrecision.Neg())


```
I want to make rounding with this library as simple as possible by not adding too many rounding methods that can cause confusion
```

Well based on the https://en.wikipedia.org/wiki/Rounding#Types_of_rounding we have 4 types of rounding. You can implement only half. Based on wiki ... round away from zero (or round toward infinity) is what you want to additionally implement for that reason RoundToInf could indeed be ok name without too much additional confusion. 

I think we can implement in some util code two with combination of Trunc and RoundToInf using them in opposite order based on the sign without the performance loss.

**Comment by quagmt** [comment]
@matejsp I added `PowToIntPart` method to simplify the needs of raising power to an exponent when the exponent is stored in Decimal (details in https://github.com/quagmt/udecimal/pull/28). That method and other methods like `Mod` and `RoundAwayFromZero` will be included in the next release

**Comment by matejsp** [comment]
That's great! I will check and adapt our wrapper to use new functions. This would simplify it a lot.

Regarding IntPart ... I still see some value with returning whole number for example we do:
```
priceTicks := r.Price.Mul(decimal.NewFromInt(100000000)).IntPart()
```
And then we call some external service with that information. For now we do String() and split fixed and decimal part. But it is slow :D

**Comment by quagmt** [comment]
Can you use `Trunc(0)`? Or you need the int part to be an integer?

**Comment by matejsp** [comment]
We need int64 to interop with third party libs (and our grpc api).
We currently as woraround use:
```
// IntPart returns the integer component of the decimal.
func (d Decimal) Int64Part() (int64, error) {
    return strconv.ParseInt(d.trunc(0).String(), 10, 64)
}
```

But it is not very fast.

**Comment by quagmt** [comment]
Added `Int64() (int,error)` as well. All needed features are included in [v1.6.0](https://github.com/quagmt/udecimal/releases/tag/v1.6.0)

**Comment by matejsp** [comment]
Perfect. Thank you very much. Will test 1.6.0 and report any issues.
